core/src/main/java/org/commonjava/indy/core/content/ContentMetadataGenerator.java

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright (C) 2011-2018 Red Hat, Inc. (https://github.com/Commonjava/indy)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.core.content;
+
+import org.commonjava.indy.IndyWorkflowException;
+import org.commonjava.indy.content.ContentGenerator;
+import org.commonjava.indy.content.DownloadManager;
+import org.commonjava.indy.content.StoreResource;
+import org.commonjava.indy.model.core.ArtifactStore;
+import org.commonjava.indy.model.core.Group;
+import org.commonjava.indy.model.core.StoreType;
+import org.commonjava.maven.galley.event.EventMetadata;
+import org.commonjava.maven.galley.io.checksum.TransferMetadata;
+import org.commonjava.maven.galley.model.Transfer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public class ContentMetadataGenerator
+                implements ContentGenerator
+{
+    public static final String FORCE_CHECKSUM_AND_WRITE = "force-checksum-and-write";
+
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    private static final Set<String> HANDLED_FILENAMES = Collections.unmodifiableSet( new HashSet<String>()
+    {
+        {
+            add( ".md5" );
+            add( ".sha1" );
+            add( ".sha256" );
+        }
+    } );
+
+    @Inject
+    private DefaultContentDigester contentDigester;
+
+    @Inject
+    private DownloadManager downloads;
+
+    @Override
+    public Transfer generateFileContent( final ArtifactStore store, final String path,
```

Comment:
Is this just to download / generate metadata files when the checksum is the first thing requested?

Reply:
I am sorry for forgetting some description here. 
This is to address NOS-1885 -- Allow automatic generation of checksums also for hosted repos. 
The word "metadata" is misleading. It means content metadata, e.g, the checksum files, and probably the http-metadata (not in this case). I borrow this concept from Galley. 
Briefly, if there is any file in hosted repo missing checksum, this will generate the checksum for them once requested. 
I added the ContentMetaGenerator and found the usage of Generators scattered in different places. So I add a manager for them to align their behavior. One major change: in the past if the canProcess returns true, but the generate fails (return null), we ignore the other generators. In this pr, I don't only rely on canProcess but also the return value of generateFileContent. If this method returns null, we keep trying the rest of Generators. 
This aligns with what we do for the other Instance<Something>, and make a lot of sense in this case. 
As you mentioned, the maven-metadata-generator already handles the .md5/sha generation, but only for maven-metadata.xml. The new generator handles the checksum for .pom/.jar/etc.

Reply:
ok, that makes sense.